### PR TITLE
Add blog article about local Kubernetes development tools

### PR DESCRIPTION
**Pull request by eyalb181**
As suggested by @sftim in another PR, this blog post offers a brief comparison between three prominent tools for local Kubernetes development: Telepresence, Gefyra, and mirrord (of which I am a maintainer). We think the methodology of connecting local apps to the Kubernetes cluster will become more prevalent in the future, and that an overview of some of the current tools could be helpful to readers of the blog.

### CODE REVIEW

File: content/en/blog/_posts/comparing-local-kubernetes-development-tools.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Comparing Local Kubernetes Development Tools: Telepresence, Gefyra, and mirrord
```
Comment (sftim) [inline content/en/blog/_posts/comparing-local-kubernetes-development-tools.md:1]: Please add valid blog article front matter (have a look at published blog articles for an example).

File: content/en/blog/_posts/comparing-local-kubernetes-development-tools.md
Comment (sftim) [inline content/en/blog/_posts/comparing-local-kubernetes-development-tools.md:1]: Please add a date into this path - see existing articles to see how we usually do it. I suggest publication on a weekday mid September, as we have a release coming up soon.

**Comment by eyalb181** [comment]
@sftim Done

**Comment by sftim** [comment]
/retitle Add blog article about local Kubernetes development tools

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+Among the various solutions, a category we dubbed “Local K8S Development tools” has emerged, which seeks to enhance the Kubernetes development experience by connecting locally running components to the Kubernetes cluster. This facilitates rapid testing of new code in cloud conditions, circumventing the traditional cycle of Dockerization, CI, and deployment.
+
+In this post, we compare three solutions in this category: Telepresence, Gefyra, and our own contender, mirrord.
+
+## Telepresence
+The oldest and most well-established solution in the category, [Telepresence](https://www.telepresence.io/) uses a VPN (or more specifically, a TUN device) to connect the user's machine (or a locally running container) and the cluster's network. It then supports the interception of incoming traffic to a specific service in the cluster, and its redirection to a local port. The traffic being redirected can also be filtered to avoid completely disrupting the remote service. It also offers complementary features to support file access (by locally mounting a volume mounted to a pod) and importing environment variables.
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:17]: nit: `tun` (not TUN; it's not an acronym).

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+Telepresence requires the installation of a local daemon (which requires root privileges) and a Traffic Manager component on the cluster. Additionally, it runs an Agent as a sidecar on the pod to intercept the desired traffic.
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:18]: Is the local daemon on your PC or somewhere else? Readers might get the wrong idea.

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+
+## Gefyra
+[Gefyra](https://gefyra.dev/), similar to Telepresence, employs a VPN to connect to the cluster but operates at the Docker network level, thus only supporting containers. This approach enhances portability across different OSes and local setups. However, the downside is that it does not support natively run uncontainerized code.
+
+Gefyra primarily focuses on network traffic, leaving file access and environment variables unsupported. Unlike Telepresence, it doesn't alter the workloads in the cluster, ensuring a straightforward clean-up process if things go awry.
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:25]: _Docker_ network? Modern clusters usually don't have a Docker network. Can you explain more about where Docker is involved and how that's relevant to people who use Kubernetes?

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+
+## mirrord
+The newest of the three tools, [mirrord](https://mirrord.dev/) adopts a different approach by injecting itself into the local binary (utilizing LD_PRELOAD or DYLD_INSERT_LIBRARIES) and overriding libc function calls, which it then proxies a temporary agent it runs in the cluster. For example, when the local process tries to read a file mirrord intercepts that call and sends it to the agent, which then reads the file from the remote pod. This method allows mirrord to cover all inputs and outputs to the process – including network access, file access, and environment variables uniformly.
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:29]:
```suggestion
The newest of the three tools, [mirrord](https://mirrord.dev/) adopts a different approach by injecting itself
into the local binary (utilizing `LD_PRELOAD` on Linux or `DYLD_INSERT_LIBRARIES` on macOS),
and overriding libc function calls, which it then proxies a temporary agent it runs in the cluster.
For example, when the local process tries to read a file mirrord intercepts that call and sends it
to the agent, which then reads the file from the remote pod. This method allows mirrord to cover
all inputs and outputs to the process – covering network access, file access, and
environment variables uniformly.
```
?

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+|  | Telepresence | Gefyra | mirrord |
+| --- | --- | --- | --- |
+| Cluster connection scope | Entire machine or container | Container | Process |
+| Supported OS | MacOS, Linux, Windows | MacOS, Linux, Windows | MacOS, Linux, Windows (WSL) |
+| Incoming traffic features | Interception | Interception | Interception or mirroring |
+| File Access | Supported | Unsupported | Supported |
+| Environment Variables | Supported | Unsupported | Supported |
+| Requires local root | Yes | No | No |
+| How to use | CLI, Docker Desktop extension | CLI, Docker Desktop extension | CLI, VS Code extension, IntelliJ plugin |
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:44]: I actually recommend using a table in HTML here:
```suggestion
<table>
<caption>Comparison of Telepresence, Gefyra, and mirrord</caption>
<thead>
<tr>
<td class="empty"></td>
<th>Telepresence</th>
<th>Gefyra</th>
<th>mirrord</th>
</tr>
</thead>
<tbody>
<tr>
<th scope="row">Cluster connection scope</th>
<td>Entire machine or container</td>
<td>Container</td>
<td>Process</td>
</tr>
<tr>
<th scope="row">Developer OS support</th>
<td>Linux, macOS, Windows</td>
<td>Linux, macOS, Windows</td>
<td>Linux, macOS, Windows (WSL)</td>
</tr>
<tr>
<th scope="row">Incoming traffic features</th>
<td>Interception</td>
<td>Interception</td>
<td>Interception or mirroring</td>
</tr>
<tr>
<th scope="row">File access</th>
<td>Supported</td>
<td>Unsupported</td>
<td>Supported</td>
</tr>
<tr>
<th scope="row">Environment variables</th>
<td>Supported</td>
<td>Unsupported</td>
<td>Supported</td>
</tr>
<tr>
<th scope="row">Requires local root</th>
<td>Yes</td>
<td>No</td>
<td>No</td>
</tr>
<tr>
<th scope="row">How to use</th>
<td><ul><li>CLI</li><li>Docker Desktop extension</li></ul></td>
<td><ul><li>CLI</li><li>Docker Desktop extension</li></ul></td>
<td>CLI</li><li>Visual Studio Code extension</li><li>IntelliJ plugin</li></ul></td>
</tr>
</tbody>
</table>
```

File: content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+<th scope="row">Requires local root</th>
+<td>Yes</td>
+<td>No</td>
+<td>No</td>
+</tr>
+<tr>
+<th scope="row">How to use</th>
+<td><ul><li>CLI</li><li>Docker Desktop extension</li></ul></td>
+<td><ul><li>CLI</li><li>Docker Desktop extension</li></ul></td>
+<td>CLI</li><li>Visual Studio Code extension</li><li>IntelliJ plugin</li></ul></td>
```
Comment (sftim) [inline content/en/blog/_posts/2023-09-12-comparing-local-kubernetes-development-tools.md:93]: Oopsn
```suggestion
<td><ul><li>CLI</li><li>Visual Studio Code extension</li><li>IntelliJ plugin</li></ul></td>
```

**Comment by sftim** [comment]
@eyalb181 could you rebase against _main_? Ideally, also squash to 1 commit.

**Comment by eyalb181** [comment]
> @eyalb181 could you rebase against _main_? Ideally, also squash to 1 commit.

Done

**Comment by sftim** [comment]
Thanks

/lgtm
/approve

**Pull request merged**